### ISSUE TITLE
Wrong suggestion in cmp_owned

### ISSUE THREAD

**Issue by antoyo** [Manishearth/rust-clippy#416]
Hello.

In the following code:

```rust
struct MyType {
    field: String,
}

impl ToString for MyType {
    fn to_string(&self) -> String {
        self.field.clone()
    }
}

fn test() {
    let my = MyType { field: "test".to_owned() };
    if my.to_string() == "test" {
    }
}
```

I get the following warning from `clippy`:

```
warning: this creates an owned instance just for comparison. Consider using `my == "test"` to compare without allocation, #[warn(cmp_owned)] on by default
```

which is obviously wrong, because the suggested code results in a compiler error.

Thanks to fix this issue.

**Comment by Manishearth** [comment]
We need to ensure that the types here are some form of `str`: https://github.com/Manishearth/rust-clippy/blob/567d5a7293f2a3ab6672919fb0ce8f94c28a6136/src/misc.rs#L179

**Comment by llogiq** [comment]
Or `Deref` into something that is `PartialEq<&str>`.

I also think in the given case (string wrapper) we should suggest adding either `Deref` or `PartialEq` to make it work.

**Comment by antoyo** [comment]
@llogiq 
What about a type that I did not created? In this case, I could not add the implementation, right?

**Comment by llogiq** [comment]
@antoyo Since your type wraps a `String` it would be easiest to just `impl Deref<String> for Mytype` where `deref(&self)` returns `&self.field`.

**Comment by Manishearth** [comment]
That's not the intended use for Deref, though.

The convert traits make sense here.

Still, this is a false positive.

**Comment by antoyo** [comment]
@llogiq 

That was just an example. Consider the following code:

```rust
fn test() {
    let number = 42;
    if number.to_string() == "42" {
    }
}
```

I get the following message from `clippy`:

```
warning: this creates an owned instance just for comparison. Consider using `number == "42"` to compare without allocation, #[warn(cmp_owned)] on by default
```

which is wrong because the code from the suggestion does not compile.
Moreover, I cannot implement the `Deref` trait on type `i32` since it is not a type defined in my crate.

**Comment by llogiq** [comment]
Agreed, this *is* a false positive. I suggest the following solutions (in increasing order of complexity):

1. Restrict the lint to `&str` types
2. Restrict the lint to types which implement `PartialEq<&str>` / `PartialOrd<&str>` (depending on comparison operator)
3. Extend the lint to include suggesting the best conversion
